applications/views.py

```diff
@@ -105,11 +106,11 @@ def applications_csv(request, city):
 
 
 @organiser_only
-def application_detail(request, city, app_id):
+def application_detail(request, city, app_number):
     """
     Display the details of a single application.
     """
-    application = Application.objects.get(pk=app_id)
+    application = [app for app in Application.objects.filter(number=app_number) if app.city == city][0]
```

Comment:
I would suggest doing something like:
`Application.objects.get(number=app_number, form__page__event__city=city)`. Here you are iterating through all the applications retrieved from database and it will get extremely slow when we will have increasing number of applications and events, also we can just retrieve the very specific application from database with one query without the need for itarating in python. Additionally iterating through applications and calling app.city on every app object will produce a database query for each of them and we don't want that :)

I know this is some tricky stuff, you can catch me on slack if you would want to discuss it further ;)